tensorflow/contrib/s3/s3_file_system.cc

```diff
@@ -0,0 +1,578 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/contrib/s3/s3_file_system.h"
+#include "tensorflow/contrib/s3/s3_crypto.h"
+#include "tensorflow/core/lib/io/path.h"
+
+#include <aws/core/Aws.h>
+#include <aws/core/utils/FileSystemUtils.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/S3Errors.h>
+#include <aws/s3/model/CopyObjectRequest.h>
+#include <aws/s3/model/DeleteObjectRequest.h>
+#include <aws/s3/model/GetObjectRequest.h>
+#include <aws/s3/model/HeadBucketRequest.h>
+#include <aws/s3/model/HeadObjectRequest.h>
+#include <aws/s3/model/ListObjectsRequest.h>
+#include <aws/s3/model/PutObjectRequest.h>
+
+#include <cstdlib>
+#include <mutex>
+
+namespace tensorflow {
+
+static const char* S3FileSystemAllocationTag = "S3FileSystemAllocation";
+static const size_t S3ReadAppendableFileBufferSize = 1024 * 1024;
+static const int S3GetChildrenMaxKeys = 100;
+
+Aws::Client::ClientConfiguration& GetDefaultClientConfig() {
+  static std::mutex cfg_lock;
+  static bool init(false);
+  static Aws::Client::ClientConfiguration cfg;
+
+  std::lock_guard<std::mutex> lock(cfg_lock);
+
+  if (!init) {
+    const char* endpoint = getenv("TFS3_ENDPOINT");
```

Comment:
Here and for the other environment variables: can we use just S3_ENDPOINT to match the conventions in tensorflow/core/platform/cloud?